### PR TITLE
gui: grey out used address in address book

### PR DESCRIPTION
**Pull request by za-kk**
Implements issue #17174, to grey out used addresses in the address book when the wallet has the `avoid_reuse` flag set.

this commit brings the `IsUsedDestination` method into the `Wallet.h` interface. It is then called in `addresstablemodel.cpp` to determine whether the address has been used or not whilst setting the font colour

master
Screenshot 2019-11-02 at 17 36 51

pr
Screenshot 2019-11-02 at 17 33 41

### CODE REVIEW

File: src/interfaces/wallet.cpp
Comment (promag) [inline src/interfaces/wallet.cpp]: There was some discussion elsewhere which, IIRC, say you should use `isUsedDestination` instead.

File: src/interfaces/wallet.h
Comment (promag) [inline src/interfaces/wallet.h]: The comment is not clear to me, what you mean by "this"?

File: src/interfaces/wallet.cpp
Comment (promag) [inline src/interfaces/wallet.cpp]: Why are you locking this?

File: src/qt/addresstablemodel.cpp
Comment (promag) [inline src/qt/addresstablemodel.cpp]: nit, here and below `if (...) {`.

File: src/qt/addresstablemodel.cpp
Comment (promag) [inline src/qt/addresstablemodel.cpp]: nit, space after `if`.

**Comment by luke-jr** [comment]
Addresses are used when they receive. Not when the UTXO created is later spent.

File: src/interfaces/wallet.h
Comment (za-kk) [reply src/interfaces/wallet.h]: this comment was carried over from the original `IsUsedDestination` method in the `wallet/wallet.h` file

File: src/interfaces/wallet.cpp
Comment (za-kk) [reply src/interfaces/wallet.cpp]: yes, makes sense as it follows the same convention as the rest of the file. will update along with other nits

File: src/interfaces/wallet.cpp
Comment (za-kk) [reply src/interfaces/wallet.cpp]: sorry, think i misunderstood what this line did, will remove

File: src/interfaces/wallet.cpp
Comment (za-kk) [reply src/interfaces/wallet.cpp]: fixed

File: src/qt/addresstablemodel.cpp
Comment (za-kk) [reply src/qt/addresstablemodel.cpp]: fixed

File: src/qt/addresstablemodel.cpp
Comment (za-kk) [reply src/qt/addresstablemodel.cpp]: fixed

File: src/interfaces/wallet.cpp
Comment (za-kk) [reply src/interfaces/wallet.cpp]: this has now been removed

**Comment by za-kk** [comment]
> Addresses are used when they receive. Not when the UTXO created is later spent.

Understood, I can update it to grey out just addresses that have received (Used) and not ones that have received and been spent if desired?

**Comment by luke-jr** [comment]
> Understood, I can update it to grey out just addresses that have received (Used) and not ones that have received and been spent if desired?

Yes (it would be meaningless/erroneous to grey out addresses based on actions of the UTXOs created when they receive).

**Comment by jonasschnelli** [comment]
Agree with @luke-jr. This is currently conceptual wrong.

**Comment by MarcoFalke** [comment]
So if the existing used-flag is supposed to do what it does (indicate true when the coin sent to that destination has been spent), it could make sense to add a new flag (indicate true when a tx has been received that has the destination in one of its outputs).

**Comment by luke-jr** [comment]
See also #15987, but I'm not sure it's the best approach to use here. If we're going to check *every address* (as this does), we should probably save a flag in the wallet/address book when we see an address used...

**Comment by za-kk** [comment]
Revisiting this pr...

I have now re-implemented by saving a flag in the address book (`address_used`) when we see an address used. I took the same approach as the `IsUsedDestination` methods and therefor this will only work when the `avoid_reuse` flag is set on the wallet.

This essentially checks the outputs when a transaction is added to the wallet (`AddToWallet`) and if it belongs to us, sets a flag in the address book (`SetUsedAddressState`). It will then grey out any addresses with this flag set when viewing them in the 'Receiving Addresses' dialog box.

**Comment by dannmat** [comment]
ack

**Comment by luke-jr** [comment]
IMO this should work for all wallets, and when loading an old wallet (or one that has been loaded by an older version since the last use of a supporting version), should populate the flag as necessary.

**Comment by za-kk** [comment]
> IMO this should work for all wallets, and when loading an old wallet (or one that has been loaded by an older version since the last use of a supporting version), should populate the flag as necessary.

@luke-jr I would be inclined to agree with you on this, it was only initially constrained to `avoid_reuse` wallets as `IsUsedDestination` was used in the first integration (although wrongly used due to confusion). So pivoting back to working for all wallets would make sense.

Would like to get others opinions on this before progressing with making the changes, thoughts @MarcoFalke @jonasschnelli @promag?

**Comment by za-kk** [comment]
Rebased with master and squashed commits.

Thinking of picking this back up if it is still a desirable change/feature? Currently it's limited to watch only wallets but as @luke-jr mentioned, this may be desirable to all wallet types

**Comment by jonasschnelli** [comment]
Concept ACK on how this works today. Slightly reviewed the code and it looks good.

Can you please review @luke-jr, @promag?

**Comment by MarcoFalke** [comment]
Open-Close to re-run ci. See https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-487591130

File: src/wallet/wallet.cpp
Comment (promag) [inline src/wallet/wallet.cpp:791]: Please update code format accordingly to developer notes.

File: src/wallet/wallet.cpp
Comment (promag) [inline src/wallet/wallet.cpp:805]: Is it me or `IsMine` is irrelevant?

Either way call `GetDestData` since it's faster and most likely to fail.

**Review by promag — COMMENTED** [review]
Maybe add a tooltip on the greyed rows?

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: Yes `IsMine` would be irrelevant in this case, have now removed to just return the output of `GetDestData` as suggested, Thanks!

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: Updated according to developer notes (added space after `if`), hopefully I haven't missed anything, Thanks!

**Comment by za-kk** [comment]
> Maybe add a tooltip on the greyed rows?

Added tooltip as suggested, not sure on suitable wording so suggestions welcome (I went with "This address has been used" followed by the original tooltip in brackets (screenshot below).

address-tooltip

**Comment by jonatack** [comment]
Concept ACK, will review.

File: src/wallet/wallet.cpp
Comment (jonatack) [reply src/wallet/wallet.cpp]: If helpful, you can use Clang in the future to check for some things:

    Use Clang formatting to clean up a file:
    clang-format -i <file>

    To format the last commit with 0 lines of context, call the
    clang-format-diff.py script from root as follows:
    git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v

File: src/wallet/wallet.cpp
Comment (jonatack) [inline src/wallet/wallet.cpp:800]: perhaps simplify
```diff
 {
     AssertLockHeld(cs_wallet);
     CTxDestination dst;
-    if (ExtractDestination(cout.scriptPubKey, dst)) {
-        if (IsMine(dst)) {
-            if (used && !GetDestData(dst, "address_used", nullptr)) {
-                AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
-            } else if (!used && GetDestData(dst, "address_used", nullptr)) {
-                EraseDestData(batch, dst, "address_used");
-            }
+    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
+        if (used && !GetDestData(dst, "address_used", nullptr)) {
+            AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
+        } else if (!used && GetDestData(dst, "address_used", nullptr)) {
+            EraseDestData(batch, dst, "address_used");
         }
     }
 }
```

File: src/qt/addresstablemodel.cpp
Comment (jonatack) [inline src/qt/addresstablemodel.cpp:233]: Suggestion: perhaps `gray` to be more compatible with dark mode.

gray:
Screenshot from 2020-06-07 10-47-18

light gray:
Screenshot from 2020-06-07 10-41-39

File: src/wallet/wallet.cpp
Comment (jonatack) [inline src/wallet/wallet.cpp:804]: Is this lock needed?

**Review by jonatack — COMMENTED** [review]
Concept ACK -- nice!

Builds cleanly and works as advertised, including the tooltip.

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: Thanks @jonatack, that will definitely be useful for me to use in the future!

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp:798]: Yes makes sense, this has now been simplified as suggested

File: src/qt/addresstablemodel.cpp
Comment (za-kk) [reply src/qt/addresstablemodel.cpp]: Thanks for testing with dark mode! have now updated the code to use `gray` instead of `lightGray` to accommodate both light and dark modes

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: Probably not needed as `GetDestData` is just accessing the address book data. I have now removed the lock.

**Comment by za-kk** [comment]
Thanks for reviewing the code @jonatack! I have now made the changes suggested and squashed my commits.

**Comment by jonatack** [comment]
Thanks for the color change. I only use the GUI in dark mode :). I don't think you needed to squash the two commits and `git diff 94cbeaa 481d3df` would have been cleaner and easier to re-review by not squashing. But wait until someone else reviews before re-touching.

File: src/wallet/wallet.cpp
Comment (jonatack) [reply src/wallet/wallet.cpp]: Sorry about that; it appears to be needed.
```
wallet/wallet.cpp:803:12: error: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
    return GetDestData(dst, "address_used", nullptr);
           ^
1 error generated.
```

**Comment by promag** [comment]
I'm sorry to NACK current approach. You shouldn't be adding blocking calls in `AddressTableModel::data`, these will make the GUI unresponsive. I think you should add a `is_used` member to `AddressTableEntry` and keep it updated accordingly.

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp:802]: My bad, should have checked properly myself before removing! This has now been updated to include the lock again.

**Comment by za-kk** [comment]
Thanks for reviewing @promag! makes sense to avoid locking up the GUI wherever possible, I will look into updating the code to follow the process you mentioned instead.

**Comment by hebasto** [comment]
Concept ACK.

@za-kk Are you still working on this? Are you going to implement @promag's [suggestion](https://github.com/bitcoin/bitcoin/pull/17355#issuecomment-640480797)?

**Comment by za-kk** [comment]
Thanks for the ACK @hebasto, yes I am still working on this and plan to implement the suggestions made by @promag soon 👍🏻

**Comment by jonatack** [comment]
Hi @za-kk, let us know if you're having any issue with progress here :rocket:

**Comment by jarolrod** [comment]
@za-kk  are you going to implement promags suggestion, or can I pick this up?

**Comment by hebasto** [comment]
> @za-kk are you going to implement promags suggestion, or can I pick this up?

Probably, the better approach is to split this change into two PRs:
- introduce new `CWallet` member functions and _add tests_ for them (in this repo)
- GUI changes (in the gui repo)

**Comment by za-kk** [comment]
Sorry, this PR has been quite stale from me over the last several months as I haven't had the chance to take a look yet. I do still plan on implementing the suggestions over the next week or so but always willing to accept help on this!

Thanks @hebasto, I was wondering what the preferred way of doing this would be now that we have the GUI repo. I will look into introducing the initial `CWallet` functions over the coming days

**Comment by promag** [comment]
@hebasto I don't quite agree with that approach. The goal is to add a GUI change and that PR should be in gui repo. If that PR requires `CWallet` changes then I think they belong there. It might make sense to do what you suggest but not in this early stage - too much effort to sync both PR's, to have split discussions etc.

**Comment by hebasto** [comment]
@promag 
> @hebasto I don't quite agree with that approach. The goal is to add a GUI change and that PR should be in gui repo. If that PR requires `CWallet` changes then I think they belong there. It might make sense to do what you suggest but not in this early stage - too much effort to sync both PR's, to have split discussions etc.

I'm fine with keeping all changes and discussion here.

New introduced functions still require test coverage (probably, in a follow up).

File: src/wallet/wallet.cpp
Comment (adamjonas) [inline src/wallet/wallet.cpp:847]: I'm getting compiler warnings consistent with the CI:

```
wallet/wallet.cpp:847:5: warning: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::__1::recursive_mutex> >' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
    AssertLockHeld(cs_wallet);
    ^
./sync.h:81:28: note: expanded from macro 'AssertLockHeld'
#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
                           ^
wallet/wallet.cpp:849:55: warning: calling function 'IsMine' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
    if (ExtractDestination(cout.scriptPubKey, dst) && IsMine(dst)) {
                                                      ^
wallet/wallet.cpp:850:22: warning: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
        if (used && !GetDestData(dst, "address_used", nullptr)) {
                     ^
wallet/wallet.cpp:851:13: warning: calling function 'AddDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
            AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
            ^
wallet/wallet.cpp:852:29: warning: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
        } else if (!used && GetDestData(dst, "address_used", nullptr)) {
                            ^
wallet/wallet.cpp:853:13: warning: calling function 'EraseDestData' requires holding mutex 'cs_wallet' exclusively [-Wthread-safety-analysis]
            EraseDestData(batch, dst, "address_used");
            ^
6 warnings generated.
```

**Review by promag — COMMENTED** [review]
I think should do the following:
 - start by adding a new `is_used` member  in `interfaces::WalletAddress` 
 - then in `getAddresses()` (src/wallet/interfaces.cpp) `is_used` should be set - you have the wallet lock there
 - add a `is_used` member to `AddressTableEntry`
 - copy `is_used` to entry in `AddressTablePriv::refreshAddressTable`
 - ensure `is_used` is updated accordingly in `AddressTablePriv::updateEntry`
 - expose in the model (like you do in this PR) but read data only from the record.

I hope this helps.

File: src/wallet/wallet.cpp
Comment (promag) [inline src/wallet/wallet.cpp:851]: Could be save txid instead of "p" and maybe reword the key to "first_txid" so that we could show the user that transaction.

File: src/wallet/wallet.cpp
Comment (promag) [inline src/wallet/wallet.cpp:845]: `used` is always `true`, maybe drop it?

File: src/wallet/wallet.cpp
Comment (promag) [inline src/wallet/wallet.cpp:861]: An existing wallet will not have these set, ignore this fact? Or just add a release note saying it that a rescan fixes this?

**Comment by hebasto** [comment]
> Hi @za-kk, let us know if you're having any issue with progress here

**Comment by za-kk** [comment]
Hi All, I finally had chance to implement the changes suggested by @promag. I notice that the test `wallet_avoidreuse.py` is currently failing in CI, I will look into resolving this.

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp:847]: Thanks for taking a look @adamjonas, this should now be resolved in latest commit

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: This should be implemented in latest commit

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: Yes I think the best solution would be adding a release note as you suggested, I will look into doing this shortly

**Comment by hebasto** [comment]
> @za-kk Are you still working on this?

Any update?

**Comment by hebasto** [comment]
@MarcoFalke @meshcollider 

Up for grabs?

**Comment by za-kk** [comment]
Apologies for the delay on this PR, I do plan on picking it back up very shortly. 

Always willing to accept help/advice to move things forward @hebasto 👍🏻

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: This has been resolved in the latest commit, thanks!

**Comment by za-kk** [comment]
This should now be ready for review again, I have implemented the suggested changes made by @promag.

**Comment by prayank23** [comment]
Concept ACK

nit: grey if used once and red if more than 1 would have been better

Will test by tomorrow

File: src/wallet/wallet.h
Comment (jonatack) [inline src/wallet/wallet.h:459]: Here and in the definition in the .cpp file, `uint256 hash` ought to be passed by reference to const, as it is not a cheaply passed type (see usage elsewhere in the codebase). 
```suggestion
    void SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

File: src/wallet/wallet.cpp
Comment (jonatack) [inline src/wallet/wallet.cpp:863]:
```suggestion
void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, const uint256& hash)
```

**Review by jonatack — COMMENTED** [review]
Thanks for updating! This is a bit of a drive-by review as the hour is late, but I was unable to build (with clang 11 on debian) after rebasing on current master at 7a49fdc :

```rake
wallet/wallet.cpp:878:15: error: out-of-line definition of 'SetUsedAddressState' does not match any declaration in 'CWallet'
void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, const uint256& hash)
              ^~~~~~~~~~~~~~~~~~~
./wallet/wallet.h:460:70: note: type of 3rd parameter of member declaration does not match definition ('uint256' vs 'const uint256 &')
    void SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, uint256 hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
                                                                     ^
wallet/wallet.cpp:883:9: error: use of undeclared identifier 'AddDestData'
        AddDestData(batch, dst, "first_txid", hash.ToString());
        ^
wallet/wallet.cpp:884:9: error: no matching function for call to object of type 'boost::signals2::signal<void (const CTxDestination &, const std::string &, bool, const std::string &, ChangeType)>' (aka 'signal<void (const variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown> &, const basic_string<char> &, bool, const basic_string<char> &, ChangeType)>')
        NotifyAddressBookChanged(this, dst, m_address_book[dst].GetLabel(), IsMine(dst) != ISMINE_NO, "receive", CT_UPDATED );
        ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/boost/signals2/detail/signal_template.hpp:720:19: note: candidate function not viable: requires 5 arguments, but 6 were provided
      result_type operator ()(BOOST_SIGNALS2_SIGNATURE_FULL_ARGS(BOOST_SIGNALS2_NUM_ARGS))
                  ^
/usr/include/boost/signals2/detail/signal_template.hpp:724:19: note: candidate function not viable: requires 5 arguments, but 6 were provided
      result_type operator ()(BOOST_SIGNALS2_SIGNATURE_FULL_ARGS(BOOST_SIGNALS2_NUM_ARGS)) const
                  ^
wallet/wallet.cpp:891:12: error: use of undeclared identifier 'GetDestData'
    return GetDestData(dst, "first_txid", nullptr);
           ^
4 errors generated.
```

File: src/wallet/wallet.h
Comment (za-kk) [reply src/wallet/wallet.h]: Thanks @jonatack, this should now be resolved in bb4790816d84f850367c99134ea6a16659fbb180

File: src/wallet/wallet.cpp
Comment (za-kk) [reply src/wallet/wallet.cpp]: Thanks @jonatack, this should now be resolved in bb4790816d84f850367c99134ea6a16659fbb180

**Comment by za-kk** [comment]
Thanks for taking a look @jonatack! seems that some latest merges to master has resulted in changes to destdata (#21353)  which was causing the build issues. I have now refactored the code to deal with this (bb4790816d84f850367c99134ea6a16659fbb180)

**Review by ghost — APPROVED** [review]
~~tACK https://github.com/bitcoin/bitcoin/pull/17355/commits/bb4790816d84f850367c99134ea6a16659fbb180~~ There are few issues which should be fixed as mentioned in https://github.com/bitcoin/bitcoin/pull/17355#pullrequestreview-699943793 and confirmed below

**Comment by prayank23** [comment]
Compiled successfully on Linux (Pop!_OS). Works as expected. Addresses were greyed out once I sent bitcoin to first 2 addresses from the 'receiving addresses':

image


Although if a user cares out privacy, should never open 'received addresses' for using because once a receiving address is created, it should be considered used. Best practice is to create a new address every time you need an address for receiving bitcoin.

File: src/wallet/wallet.cpp
Comment (jonatack) [inline src/wallet/wallet.cpp:884]:
```suggestion
        m_address_book[dst].destdata.emplace(std::make_pair("first_txid", hash.ToString()));
```

File: src/qt/addresstablemodel.cpp
Comment (jonatack) [inline src/qt/addresstablemodel.cpp:238]:
```suggestion
        if (rec->is_used) {
```

File: src/qt/addresstablemodel.cpp
Comment (jonatack) [inline src/qt/addresstablemodel.cpp:243]:
```suggestion
        if (rec->is_used) {
```

File: src/wallet/wallet.h
Comment (jonatack) [inline src/wallet/wallet.h:460]: The first argument of `SetAddressReceivedState()` seems unused unless I'm missing something.
```suggestion
    void SetAddressReceivedState(const CTxOut& cout, const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```

File: src/wallet/wallet.cpp
Comment (jonatack) [inline src/wallet/wallet.cpp:903]: style nit: clang format and also can simplify
```diff
@@ -890,16 +890,8 @@ bool CWallet::HasAddressReceived(const CTxDestination& dst) const
 {
 
     const std::string key{"first_txid"};
-    std::map<CTxDestination, CAddressBookData>::const_iterator i = m_address_book.find(dst);
-    if(i != m_address_book.end())
-    {
-        CAddressBookData::StringMap::const_iterator j = i->second.destdata.find(key);
-        if(j != i->second.destdata.end())
-        {
-            return true;
-        }
-    }
-    return false;
+    const std::map<CTxDestination, CAddressBookData>::const_iterator it = m_address_book.find(dst);
+    return it != m_address_book.end() && it->second.destdata.find(key) != it->second.destdata.end() ? true : false;
 }
```

**Review by jonatack — COMMENTED** [review]
Some comments for your consideration. The pull doesn't necessarily need to be done in one commit depending on how you organise the changes and provided they are hygienic, but the "fix up to make it build" commit probably should be squashed into the relevant changes. Don't hesitate to run clang-format on your changes.

File: src/qt/addresstablemodel.cpp
Comment (jonatack) [reply src/qt/addresstablemodel.cpp:243]: (or for both, maybe simplify further)
```diff
@@ -234,15 +234,11 @@ QVariant AddressTableModel::data(const QModelIndex &index, int role) const
         } // no default case, so the compiler can warn about missing cases
         assert(false);
     }
-    else if (role == Qt::ForegroundRole) {
-        if (rec->is_used == true) {
-            return QColor(Qt::gray);
-        }
+    else if (role == Qt::ForegroundRole && rec->is_used) {
+        return QColor(Qt::gray);
     }
-    else if (role == Qt::ToolTipRole) {
-        if (rec->is_used == true) {
-            return tr("This address has been used (Right-click to edit address or label)");
-        }
+    else if (role == Qt::ToolTipRole && rec->is_used) {
+        return tr("This address has been used (Right-click to edit address or label)");
     }
     return QVariant();
```

File: src/wallet/wallet.cpp
Comment (jonatack) [inline src/wallet/wallet.cpp:883]: nit, omit extra line break here and also line 891 below

**Review by ryanofsky — COMMENTED** [review]
Reviewing bb4790816d84f850367c99134ea6a16659fbb180. This seems pretty good but:

- It seems like `SetAddressReceivedState` should be called in `LoadToWallet` as well as `AddToWallet`, not just `AddToWallet` so old and new addresses are displayed correctly, not just ~old~new addresses.
- It seems like `SetAddressReceivedState` should be called regardless of `WALLET_FLAG_AVOID_REUSE` flag. The avoid reuse feature is for detecting addresses that we previously _spent from_  (and avoiding spending from them again). This feature is about detecting which addresses we previously _received to_, which is not directly related, and probably doesn't need a feature flag since it doesn't affect the wallet database storage.
- Would seem more efficient and simpler to a add new bool `CAddressBookData::received` member instead of adding a `"first_txid"` entry and transaction hash to the `CAddressBookData::destdata` map
- Could be nice to have some test coverage for this in https://github.com/bitcoin/bitcoin/blob/master/src/qt/test/wallettests.cpp

EDIT: corrected old/new above

**Comment by prayank23** [comment]
> It seems like SetAddressReceivedState should be called in LoadToWallet as well as AddToWallet, not just AddToWallet so old and new addresses are displayed correctly, not just old addresses.

I relaunched `bitcoin-qt` today and my old addresses which had received bitcoin, were marked [grey yesterday](https://github.com/bitcoin/bitcoin/pull/17355#issuecomment-874272237) are not greyed out now:

image

> It seems like SetAddressReceivedState should be called regardless of WALLET_FLAG_AVOID_REUSE flag. The avoid reuse feature is for detecting addresses that we previously spent from (and avoiding spending from them again). This feature is about detecting which addresses we previously received to, which is not directly related, and probably doesn't need a feature flag since it doesn't affect the wallet database storage.

Agree

**Comment by za-kk** [comment]
Thanks for the feedback everyone! Closing this PR for now until I have more time to come back and implement the changes, anyone else is welcome to pick it up in the meantime if desired 😀

**Comment by fanquake** [comment]
Removing "Up for grabs" here. Followup could start in https://github.com/bitcoin-core/gui/issues/528, or the QML repo.

**Pull request closed without merging**
